version: export tweak version in version.h and version string

tejlmand opened the pull request:
The PR exports TWEAK version to code through version.h. This aligns the symbols available in code with those available in CMake and Kconfig.

It also introduces a new extended version string containing all version info fields, including tweak.

The new define and string is available for both KERNEL, APP, and custom version types (such as MCUBOOT).

tejlmand committed:
version: export tweak version in version.h

The commit export TWEAK version to code through version.h.
This aligns the symbols available in code with those available in
CMake and Kconfig.

The new define is available for both KERNEL, APP, and custom version
types (such as MCUBOOT).

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

nordicjm reviewed (changes requested):
Doc update needed in `doc/build/version/index.rst`

tejlmand commented:
> Doc update needed in `doc/build/version/index.rst`

Fixed.

pdgendt commented:
Does the tweak version need to be added to "version binary descriptors" in `subsys/bindesc/Kconfig.version`? Can be done in a separate PR probably.

tejlmand committed:
version: cmake: kconfig: introduce extra <type>_VERSION_<x>_STRING

Fixes: #68360

This commit introduces <type>_VERSION_TWEAK_STRING which includes
the tweak field in the string, but without the extra version.
This format is used by MCUboot / imgtool, and thus makes it easier to
align code to the format used by MCUboot.

This commit also introduces <type>_VERSION_EXTENDED_STRING which
includes the tweak field in the string in addition to the extra version
field.

The new defines / variables is available in code, CMake, and Kconfig,
and it defined for KERNEL, APP, and custom types, such as MCUBOOT.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented:
> Does the tweak version need to be added to "version binary descriptors" in `subsys/bindesc/Kconfig.version`? Can be done in a separate PR probably.

Probably a good idea, especially as tweak has always been present in `VERSION` file, but as this PR only touched handling of VERSION file and making values available in various contexts, then I believe extending bindesc warrants its own PR, especially as there seems to be a bit more to it, than just adding a Kconfig setting.

kkasperczyk-no commented on doc/build/version/index.rst:
```diff
@@ -73,85 +73,105 @@ To use the version information in application code, the version file must be inc
+|                             |                 | ``VERSION_MINOR``, |br|                           |                  |
+|                             |                 | ``PATCHLEVEL``, |br|                              |                  |
+|                             |                 | ``EXTRAVERSION``, |br|                            |                  |
+|                             |                 | ``PATCHLEVEL``                                    |                  |
```
Shouldn't it be
```suggestion
|                             |                 | ``VERSION_TWEAK``                                    |                  |
```
?

tejlmand replied:
it sure should.

tejlmand replied:
fixed

LuDuda reviewed (approved):
Looks awesome 👍 

A bit unrelated, but this line does not look correct: https://github.com/zephyrproject-rtos/zephyr/pull/68160/files#diff-4919a7623a2d331ff10447d41f4d1f705af5f6962120735553687400f5bd08fbL163

nordicjm committed:
mcuboot: Use tweak version string

Replaces the default with a shorter variable of equal value

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
